README.md
# nifty-nav-2
[![Build Status](https://travis-ci.org/factor1/nifty-nav-2.svg?branch=master)](https://travis-ci.org/factor1/nifty-nav-2)
![Github file size](https://img.shields.io/github/size/factor1/nifty-nav-2/dist/niftyNav2.js.svg)
![npm](https://img.shields.io/npm/v/nifty-nav-2.svg)
![GitHub issues](https://img.shields.io/github/issues-raw/factor1/nifty-nav-2.svg)
![license](https://img.shields.io/github/license/factor1/nifty-nav-2.svg)
![David Dependencies](https://david-dm.org/factor1/nifty-nav-2.svg)

A customizable JavaScript menu framework for modern browsers made easy.

## Getting Started
### 🖥 Installation
---
Add Nifty Nav 2 to your project using your favorite package manager.

##### npm
```bash
npm install nifty-nav-2 --save
```

##### yarn
```bash
yarn add nifty-nav-2
```

###  📂 Including in your project
---
#### Styles
Import the Nifty Nav 2 css or scss, or ensure it is included in your build. The following
is using scss to import from `node_modules`. Alternatively, you could copy `niftyNav2.css`
from `node_modules` to your project.

```scss
// include nifty nav 2 styles
@import 'nifty-nav-2/dist/niftyNav2.css'
```

#### JavaScript
How you include the Nifty Nav 2 JavaScript library is up to you and the architecture
of your project. Nifty Nav 2 is available as a UMD module so it can be used as a
AMD, Global, Node, or ES6 module. The compiled JavaScript file found in `dist/` contains
the `babel-polyfill` to ensure proper cross browser support.

##### Simple
Copy the `niftyNav2.js` file in `nifty-nav-2/dist` to your project directory and include
in your HTML. You do not need to include any polyfills, as they are bundled directly
into `dist/niftyNav2.js`.

```html
<html>
  <body>
    <!-- Your HTML Content -->
    <script src="path_to_your_assets/js/niftyNav2.js"></script>
  </body>
</html>
```

##### Using Gulp
A Gulp implementation that concatenates JS files may look something like this.

```js
var concatFiles = [
  './node_modules/nifty-nav-2/dist/niftyNav2.js', // adds niftyNav2.js to be concatenated
  './assets/js/*.js', // grabs all js files from our assets/js folder
  '!./assets/js/*.min.js', // ignore minified files
]

// Concatenate & Minify JavaScript
gulp.task('scripts', ['lint'], function() {
  return gulp.src( concatFiles )
    .pipe(concat( 'all.js' ))
    .pipe(gulp.dest( './assets/js/' ))
    .pipe(rename('all.min.js'))
    .pipe(uglify())
    .pipe(gulp.dest( './assets/js/' ))
});
```

##### ES6 Module
```js
import niftyNav2 from 'nifty-nav-2';
```

### 🛠 Usage
---
#### Adding the Trigger
Place the trigger or "hamburger". You can place this wherever you want it to appear
in the DOM.

```HTML
<div data-nifty-target="niftyNav"></div>
```

Your element must have a `data-nifty-target` attribute with a unique name. Here,
we are simply calling it `niftyNav` but it can be anything you wish. If changed from
the default `niftyNav` be sure to pass that via settings when initializing.

#### Adding the Panel
Add the panel that will be opened on click. The ID should match the value of `data-nifty-target`.

```HTML
<div id="niftyNav">
  <!--Panel Contents -->
</div>
```

#### Initialization
Initialize Nifty Nav 2 via JavaScript. The following is an example with settings
for `iconColor` and `panelColor`.

```js
<script>
  niftyNav2.init({
    iconColor: 'rgb(123, 249, 158)',
    panelColor: 'rgb(123, 249, 158)',
  });
</script>
```

## ⚙️ Settings
There are plenty settings that can be used to customize Nifty Nav 2.

| Setting | Description | Accepts | Default |
|---------|-------------|---------|---------|
| `targets` | Targets to create/use Nifty Nav 2 (from `data-nifty-target`) | **array** | `['niftyNav']`
| `icon`    | Icon style to be used | **string** - square, rounded | `square` |
| `iconColor` | Color to be applied to the icon | **string** | `#fff` |
| `showMenuText` | Show or hide text next to the icon | **boolean** | `false` |
| `menuText` | Text to appear when `showMenuText` is `true` | **string** | `Menu` |
| `menuTextColor` | Color to be applied to `menuText` | **string** | `#fff` |
| `panelOrigin` | Placement of where the Nifty Nav 2 panel will animate from | **string** - top, left, right | `top` |
| `panelTopOffset` | Top offset for the panel | **integer** | `0` |
| `panelPosition` | CSS positioning for the panel | **string** | `absolute` |
| `panelHeight` | Height to be applied to the panel | **string** | `auto` |
| `panelWidth` | Width to be applied to the panel | **string** | `100%` |
| `panelAnimation` | Type of animation for the panel | **string** - slide-in, fade-in, off | `slide-in` |
| `panelAnimationSpeed` | Speed of panel animation | **integer** | `500` |
| `panelColor` | Color to be applied to the background of the panel | **string** | `#2d2d2d` |
| `showMask` | Show or hide the mask/overlay | **boolean** | `true` |
| `maskAnimationSpeed` | Speed at which the mask/overlay animates in | **integer** | `600` |
| `maskColor` | Color to be applied to the mask/overlay | **string** | `rgba(0,0,0,0.5)` |

## 🔌 API Methods
There are a few API Methods available to you if you should need them.

| Method | Description | Accepts | Default |
|--------|-------------|---------|---------|
| `onToggle` | Accepts a function that runs when `togglePanel` is fired | x | x |
| `toggleIcons` | Toggles the state of all icons | x | x |
| `toggleMask` | Toggles the mask when used | x | x |
| `togglePanel` | Toggles the panel when used | `panelId` which is generated from the `targets` options | none |

## 🌎 Browser Support
Current browser support as of **v0.1.0**.

| Browser | Support | Notes |
|---------|:-------:|-------|
| Edge 16+ | ✅ |       |
| Edge <= 15 | ✅ |  |
| IE 11 | ✅ |  |
| IE 10 | ⚠️ | May need CSS to adjust hamburger and menu text placement, core functionality is OK |
| Firefox 51+ | ✅ |  |
| Chrome 51+ | ✅ |   |
| Safari 9.1+ | ✅ |  |


## ⚠️ Issues & Contributions
Feel free to open an issue or a pull request if you find any issues or enhancements.

## 📝 Changelog
The changelog can be found [here](https://github.com/factor1/nifty-nav-2/blob/master/CHANGELOG.md).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.3.2] - 2018-06-27
### Fix
- adds CSS to avoid a bug where the mask could be double clicked (Issue #6)

### Adds
- `toggleIcon` API method to toggle any and all icon states

## [0.3.1] - 2018-06-25
### Fix
- adding a `panelHeight` setting during `init()` now works instead of doing nothing

## [0.3.0] - 2018-06-25
### Adds
- `onToggle` method as a callback for `togglePanel`

### Updates
- Readme info for API methods  

## [0.2.0] - 2018-04-12
### Updates
- How JS styles are applied to elements
- Update icon structure to be more cross browser friendly

### Adds
- `babel-polyfill` to dist bundle.

## [0.1.0] - 2018-03-28
### Added
- Nifty Nav 2 was born!


[0.3.1]: https://github.com/factor1/nifty-nav-2/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/factor1/nifty-nav-2/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/factor1/nifty-nav-2/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/factor1/nifty-nav-2/compare/v0.0.2-0...v0.1.0


webpack.config.js
const webpack = require('webpack');
const autoprefixer = require('autoprefixer');
const path = require('path');
const ExtractTextPlugin = require('extract-text-webpack-plugin');
const CleanWebpackPlugin = require('clean-webpack-plugin');

module.exports = {
  entry: [
    'babel-polyfill',
    './src/js/niftyNav2.js'
  ],
  devtool: 'source-map',
  output: {
    path: path.resolve(__dirname, 'dist'),
    publicPath: path.resolve(__dirname, 'dist'),
    filename: 'niftyNav2.js',
    library: 'niftyNav2',
    libraryTarget: 'umd',
    sourceMapFilename: '[file].map'
  },
  watchOptions: {
    ignored: ['./node_modules', './dist']
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        loader: 'babel-loader'
      },
      {
        test: /\.scss$/,
        use: ExtractTextPlugin.extract({
          fallback: 'style-loader',
          use: [
            {
              loader: 'css-loader',
              options: {
                minimize: true,
                sourceMap: true
              }
            }, {
              loader: 'sass-loader',
              options: {
                sourceMap: true
              }
            }, {
              loader: 'postcss-loader',
            }
          ]
        })
      }
    ]
  },
  plugins: [
    new CleanWebpackPlugin(['dist']),
    new webpack.optimize.UglifyJsPlugin(),
    new ExtractTextPlugin('niftyNav2.css')
  ]
}


src/js/niftyNav2.js
/**
  * Nifty Nav 2
  * A JavaScript menu framework.
  * Author: Eric Stout, Factor1 Studios
  * https://github.com/factor1/nifty-nav-2
**/

// eslint-disable-next-line
import styles from '../scss/niftyNav2.scss';


/**
 * Default Options
 */
let options = {
  icon: 'square', // icon style (square, rounded)
  iconColor: '#fff', // default icon color
  maskAnimationSpeed: 600, // speed of the mask animation
  maskColor: 'rgba(0,0,0,0.5)',
  menuText: 'Menu', // text to appear when showMenuText is true
  menuTextColor: '#fff', // menu text color
  panelAnimation: 'slide-in', // type of panel animation (slide-in, fade-in, off)
  panelAnimationSpeed: 500, // speed of panel animation,
  panelColor: '#2d2d2d', // panel color
  panelHeight: 'auto', // panel height
  panelOrigin: 'top',  // where the panel will animate or originate from (top/left/right)
  panelPosition: 'absolute', // css position - absolute, relative, fixed, etc...
  panelTopOffset: 0, // top offset for the panel
  panelWidth: '100%', // panel width
  showMask: true, // if there should be a mask covering page content,
  showMenuText: false, // toggle text next to icon
  targets: ['niftyNav'] // targets can be an array so you can have multiple instances at a time <div data-nifty-target="niftyNav"></div>
}

/**
 *
 * Build Icon Structure
 *
**/
const buildIcons = (target, options) => {

  const icon = `
    <button class="nifty-button">
      <div class="nifty-button__container">
        <div class="nifty-icon nifty-icon--${options.icon}">
          <span style="background-color: ${options.iconColor}"></span>
          <span style="background-color: ${options.iconColor}"></span>
          <span style="background-color: ${options.iconColor}"></span>
        </div>
      </div>
    </button>
  `;
  target.innerHTML = icon;

  // if we are showing menu text
  if( options.showMenuText === true ) {
    target.querySelector('.nifty-button').innerHTML = `
      <div class="nifty-button__container">
        <div class="nifty-icon--text" style="color: ${options.menuTextColor};">
          <span>${options.menuText}</span>
        </div>
        <div class="nifty-icon nifty-icon--${options.icon}">
          <span style="background-color: ${options.iconColor}"></span>
          <span style="background-color: ${options.iconColor}"></span>
          <span style="background-color: ${options.iconColor}"></span>
        </div>
      </div>
    `;
  }

}

/**
 *
 * Build Panel
 *
**/
const buildPanel = (target, options) => {
  // get the panel id from the data attr
  const panelId = target.getAttribute('data-nifty-target');
  const panel = document.getElementById(panelId);

  // add nifty-panel class automatically
  panel.classList.add('nifty-panel');

  // Panel Top Offset Setting
  if( options.panelTopOffset !== 0 ) {
    panel.style.cssText += `top: ${options.panelTopOffset}px`;
  }

  // Panel Origin Setting
  panel.classList.add(`nifty-panel--${options.panelOrigin}`);

  // Panel Width Setting
  if( options.panelWidth !== '100%' ) {
    panel.style.cssText += `width: ${options.panelWidth};`;
  }

  // Panel Height Setting
  if( options.panelHeight !== 'auto' ) {
    panel.style.cssText += `height: ${options.panelHeight};`;
  }

  // Panel Color
  panel.style.cssText += `background-color: ${options.panelColor};`;

  // Panel Position Setting
  panel.style.cssText += `position: ${options.panelPosition};`

  // Panel Animation Speed Setting
  panel.style.cssText += `transition: all ${options.panelAnimationSpeed}ms ease-in-out;`;

  // Set Panel State to Closed
  panel.classList.add('nifty-panel--closed');

  // handle animation options
  if( options.panelAnimation === 'fade-in' ) {
    panel.classList.add('nifty-panel--fade-in');
  } else if ( options.panelAnimation === 'off') {
    panel.classList.add('nifty-panel--off');
  }

}

/**
 *
 * On Toggle - Fires a callback function when passed to togglePanel()
 *
**/
let onToggleCallback = null;
const onToggle = ( callback ) => {
  if( callback ) {
    onToggleCallback = callback;
  }
};

/**
 *
 * Open/Close Panel
 *
**/
const togglePanel = (panelId, callback = onToggleCallback) => {
  const panel = document.getElementById(panelId);

  panel.classList.toggle('nifty-panel--open');

  if( callback ) {
    return callback();
  }
}

/**
 *
 * Toggle Mask
 *
**/
const toggleMask = () => {
  const mask = document.getElementById('niftyMask');

  if( mask.classList.contains('nifty-mask--active') ) {

    mask.classList.remove('nifty-mask--active');
    mask.classList.add('nifty-mask--closing');

    setTimeout( () => {
      mask.classList.remove('nifty-mask--closing')
    }, 500);

  } else {
    mask.classList.toggle('nifty-mask--active');
  }

}

/**
 *
 * Add Mask to DOM
 *
**/
const addMask = () => {
  // setup element
  const mask = document.createElement('div');
  mask.setAttribute('id', 'niftyMask');
  mask.setAttribute('class', 'nifty-mask');
  mask.style.cssText += `background: ${options.maskColor}`;

  // if animations are disabled set transition to none
  if( options.panelAnimation === 'off' ) {
    mask.style.cssText += 'transition: none;';
  } else {
    // set the transition for the mask
    mask.style.cssText += `opacity: ${options.maskAnimationSpeed}ms ease-in-out`;
  }

  // showMask setting
  if( !options.showMask ) {
    mask.style.cssText += 'opacity: 0;';
  }

  // add click listener
  mask.addEventListener('click', () => {

    toggleMask();

    // shut down all open nifty nav panels
    const panels = document.querySelectorAll('.nifty-panel');

    for(let i = 0; panels.length > i; i++ ) {
      panels[i].classList.remove('nifty-panel--open');
    }

    // remove all active icons
    const icons = document.querySelectorAll('.nifty-icon');

    for(let i = 0; icons.length > i; i++ ) {
      icons[i].classList.remove('nifty-active');
    }

  });

  return document.body.appendChild(mask);
}

/**
 *
 * Handle click(s) of target listener
 *
**/
const handleTargetClick = (e) => {
  const panelId = e.target.parentElement.getAttribute('data-nifty-target');

  const icon = e.target.querySelector('.nifty-icon');
  icon.classList.toggle('nifty-active');

  togglePanel(panelId);
  toggleMask();
}

/**
 *
 * Toggle All Icon States
 *
**/
const toggleIcons = () => {
  const icons = document.querySelectorAll('.nifty-icon');

  icons.forEach( ( icon ) => {
    icon.classList.toggle('nifty-active');
  });
}

/**
 *
 * Initialization of niftyNav2
 *
**/
const init = (settings) => {
  // Detect not supported browsers (<=IE9)
  const browserNotSupported = document.all && !window.atob;

  if( browserNotSupported ) {
    return console.error('[Nifty Nav 2]: Browser not supported. Please upgrade your browser.');
  }

  // get the defaults and user settings
  options = Object.assign(options, settings);

  // handle click(s) on target
  const clickTargets = document.querySelectorAll('div[data-nifty-target]');


  for (const [i, target] of clickTargets.entries()) {
    buildIcons(target, options);
    buildPanel(target, options);
    target.addEventListener('click', (e) => {handleTargetClick(e)} );
  }

  // Add mask (even if false, as its used for click to close)
  addMask();

}

/**
 *
 * Export Public API
 *
**/
module.exports = {
  init,
  onToggle,
  toggleIcons,
  toggleMask,
  togglePanel
}
